frontend/src/Routes/DynamicRoute/DynamicRoute.tsx

```diff
@@ -16,62 +16,51 @@ type DynamicRouteProps = {
   location?: Location;
 };
 
-type RoutePage = {
-  path: string;
-  exact?: boolean;
-};
-
-const checkPath = (pathname, { path, exact }: RoutePage) => {
-  const [, section] = pathname.split('/');
-  return matchPath(pathname, { path, exact }) || matchPath(pathname, { path: `/${section}${path}`, exact });
-};
-const DynamicRoute: React.FC<DynamicRouteProps> = ({ location }) => {
-  const [Component, setComponent] = React.useState<React.ExoticComponent<any>>(React.Fragment);
-  const [currClassName, setCurrClassName] = React.useState<string>(null);
-  // TODO: This seems to get executed when the tree fails to have PluginStoreProvider and then explodes
+type StateRoute = {
+  className: string;
+  Component: React.ExoticComponent<any>;
+}
+const DynamicRoute: React.FC<DynamicRouteProps> = () => {
+  const [routes, setRoutes] = React.useState<StateRoute[]>([]);
   const dynamicRoutePages = useExtensions<DynamicRoutePage>(isDynamicRoutePage);
+
   React.useEffect(() => {
-    if (location) {
-      const { properties: currRoute, pluginName } =
-        dynamicRoutePages.find(({ properties }) => {
-          if (Array.isArray(properties.path)) {
-            return properties.path.some((path) => checkPath(location.pathname, { ...properties, path }));
+    setRoutes(dynamicRoutePages.map(({ properties: { component, ...currRoute }, pluginName }) => {
```

Comment:
use `useMemo` and remove the `routes` state.

Reply:
I tent to avoid `useMemo`, but it's a good fit for this one. What is your opinion on `useEffect` + `useRef`, it'll basically do the same thing but without the pitfall of React ignoring some memoized values as mantioned in the docs[1]

[1] https://reactjs.org/docs/hooks-reference.html#usememo

Reply:
always `useMemo` when you want the value to be computed in the render pass. `useEffect` is run _after_ render and therefore should only be used when computing a side effect or something that requires async computation. 

`useEffect` + `useState` will result in 2 render passes vs `useMemo`.
`useEffect` + `useRef` would be a special case because updating the ref doesn't cause a re-render at all.

Reply:
Also, if the code can be written entirely in the render without any hook, then `useMemo` is the appropriate choice because it's only there to improve performance.